rcljava/src/main/java/org/ros2/rcljava/action/ActionServerImpl.java

```diff
@@ -0,0 +1,561 @@
+    private static native void nativeGoalEventExecute(long goalHandle);
+    private static native void nativeGoalEventCancelGoal(long goalHandle);
+    private static native void nativeGoalEventSucceed(long goalHandle);
+    private static native void nativeGoalEventAbort(long goalHandle);
+    private static native void nativeGoalEventCanceled(long goalHandle);
+    private static native void nativeDispose(long handle);
+
+    public GoalHandleImpl(
+      ActionServer<T> actionServer, action_msgs.msg.GoalInfo goalInfo, MessageDefinition goal)
+    {
+      this.actionServer = actionServer;
+      this.goalInfo = goalInfo;
+      this.goal = goal;
+      long goalInfoFromJavaConverterHandle = goalInfo.getFromJavaConverterInstance();
+      long goalInfoDestructorHandle = goalInfo.getDestructorInstance();
+      this.handle = nativeAcceptNewGoal(
+        actionServer.getHandle(),
+        goalInfoFromJavaConverterHandle,
+        goalInfoDestructorHandle,
+        goalInfo);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized action_msgs.msg.GoalInfo getGoalInfo() {
+      return this.goalInfo;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized MessageDefinition getGoal() {
+      return this.goal;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized GoalStatus getGoalStatus() {
+      int status = nativeGetStatus(this.handle);
+      return GoalStatus.fromMessageValue((byte)status);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized boolean isCanceling() {
+      return this.getGoalStatus() == GoalStatus.CANCELING;
+    }
+
+    /**
+     * Transition the goal to the EXECUTING state.
+     */
+    public synchronized void execute() {
+      // It's possible that there has been a request to cancel the goal prior to executing.
+      // In this case we want to avoid the illegal state transition to EXECUTING
+      // but still call the users execute callback to let them handle canceling the goal.
+      if (!this.isCanceling()) {
+        nativeGoalEventExecute(this.handle);
+      }
+    }
+
+    /**
+     * Transition the goal to the CANCELING state.
+     */
+    public synchronized void cancelGoal() {
+      nativeGoalEventCancelGoal(this.handle);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized void succeed() {
+      nativeGoalEventSucceed(this.handle);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized void canceled() {
+      nativeGoalEventCanceled(this.handle);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized void abort() {
+      nativeGoalEventAbort(this.handle);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized final void dispose() {
+      nativeDispose(this.handle);
+      this.handle = 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public synchronized final long getHandle() {
```

Comment:
Depending how this function is used, there's a possible race with `dispose()`. At second glance, I'm not sure we need this particular method. If possible I'll just remove it entirely. I'll wait  to see what things look like after completing the implementation.

Reply:
The problem is that, synchronized or not, there's still a race with dispose.
e.g.:
- Thread A calls dispose
- Thread B gets the handle (synchronized with dispose) and use it afterwards (no synchronized with dispose).

this can already happen in the executor, which "borrows" the handles but cannot synchronize with dispose.

> If possible I'll just remove it entirely

IIUC, this is part of the `Disposable` interface and needed by the executor.
We should probably improve it later, and instead of returning a `long` return a `Handle` object that avoids this problem.

Reply:
> IIUC, this is part of the Disposable interface and needed by the executor.

This method is in `GoalHandleImpl` and not called by the executor. So, we might not need it.
I've removed the synchronized keyword for now so there's no assumption that it is thread-safe: 36353b8